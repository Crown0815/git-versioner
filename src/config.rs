use anyhow::anyhow;
use clap::Parser;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

pub const MAIN_BRANCH: &str = r"^(trunk|main|master)$";
pub const RELEASE_BRANCH: &str = r"^releases?[/-](?<BranchName>.+)$";
pub const FEATURE_BRANCH: &str = r"^features?[/-](?<BranchName>.+)$";
pub const VERSION_PATTERN: &str = r"^[vV]?(?<Version>.+)";
pub const PRERELEASE_TAG: &str = "pre";

pub trait Configuration {
    fn repository_path(&self) -> &PathBuf;
    fn main_branch(&self) -> &str;
    fn release_branch(&self) -> &str;
    fn feature_branch(&self) -> &str;
    fn version_pattern(&self) -> &str;
    fn prerelease_tag(&self) -> &str;
    fn verbose(&self) -> bool {
        false
    }
}

#[derive(Debug)]
pub struct DefaultConfig {
    pub path: PathBuf,
    pub main_branch: String,
    pub release_branch: String,
    pub feature_branch: String,
    pub version_pattern: String,
    pub prerelease_tag: String,
}

#[derive(Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct ConfigurationFile {
    pub main_branch: Option<String>,
    pub release_branch: Option<String>,
    pub feature_branch: Option<String>,
    pub version_pattern: Option<String>,
    pub prerelease_tag: Option<String>,
}

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
pub struct Args {
    #[arg(short, long, value_parser)]
    path: Option<PathBuf>,

    #[arg(long, value_parser)]
    main_branch: Option<String>,

    #[arg(long, value_parser)]
    release_branch: Option<String>,

    #[arg(long, value_parser)]
    feature_branch: Option<String>,

    #[arg(long, value_parser)]
    version_pattern: Option<String>,

    #[arg(long, value_parser)]
    prerelease_tag: Option<String>,

    #[arg(short, long)]
    verbose: bool,

    /// Path to a configuration file (TOML or YAML)
    #[arg(short = 'c', long = "config")]
    config_file: Option<PathBuf>,
}

#[derive(Debug)]
pub struct ConfigurationLayers {
    args: Args,
    file: ConfigurationFile,
    config: DefaultConfig,
}

impl Default for DefaultConfig {
    fn default() -> Self {
        Self {
            path: ".".into(),
            main_branch: MAIN_BRANCH.to_string(),
            release_branch: RELEASE_BRANCH.to_string(),
            feature_branch: FEATURE_BRANCH.to_string(),
            version_pattern: VERSION_PATTERN.to_string(),
            prerelease_tag: PRERELEASE_TAG.to_string(),
        }
    }
}

impl Configuration for DefaultConfig {
    fn repository_path(&self) -> &PathBuf {
        &self.path
    }
    fn main_branch(&self) -> &str {
        &self.main_branch
    }
    fn release_branch(&self) -> &str {
        &self.release_branch
    }
    fn feature_branch(&self) -> &str {
        &self.feature_branch
    }
    fn version_pattern(&self) -> &str {
        &self.version_pattern
    }
    fn prerelease_tag(&self) -> &str {
        &self.prerelease_tag
    }
}

impl ConfigurationFile {
    pub fn from_file<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
        let path = path.as_ref();
        let extension = path
            .extension()
            .and_then(|ext| ext.to_str())
            .ok_or_else(|| anyhow!("File has no extension"))?;

        match extension.to_lowercase().as_str() {
            "toml" => Self::from_toml_file(path),
            "yaml" | "yml" => Self::from_yaml_file(path),
            _ => Err(anyhow!("Unsupported file format: {}", extension)),
        }
    }

    pub fn from_toml_file<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
        let content = fs::read_to_string(path)?;
        let config: Self = toml::from_str(&content)?;
        Ok(config)
    }

    pub fn from_yaml_file<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
        let content = fs::read_to_string(path)?;
        let config: Self = serde_yaml::from_str(&content)?;
        Ok(config)
    }

    pub fn from_default_files() -> anyhow::Result<Self> {
        let toml_path = Path::new(".git-versioner.toml");
        let yaml_path = Path::new(".git-versioner.yaml");
        let yml_path = Path::new(".git-versioner.yml");

        if toml_path.exists() {
            return Self::from_toml_file(toml_path);
        } else if yaml_path.exists() {
            return Self::from_yaml_file(yaml_path);
        } else if yml_path.exists() {
            return Self::from_yaml_file(yml_path);
        }

        Err(anyhow!("No configuration file found"))
    }
}

pub fn load_configuration() -> anyhow::Result<ConfigurationLayers> {
    let args = Args::parse();
    let config = DefaultConfig::default();
    let file = match &args.config_file {
        None => ConfigurationFile::from_default_files(),
        Some(path) => ConfigurationFile::from_file(path),
    }
    .unwrap_or_default();
    Ok(ConfigurationLayers { args, file, config })
}

impl Configuration for ConfigurationLayers {
    fn repository_path(&self) -> &PathBuf {
        if let Some(path) = &self.args.path {
            path
        } else {
            &self.config.path
        }
    }
    fn main_branch(&self) -> &str {
        if let Some(branch) = &self.args.main_branch {
            branch
        } else if let Some(branch) = &self.file.main_branch {
            branch
        } else {
            &self.config.main_branch
        }
    }
    fn release_branch(&self) -> &str {
        if let Some(branch) = &self.args.release_branch {
            branch
        } else if let Some(branch) = &self.file.release_branch {
            branch
        } else {
            &self.config.release_branch
        }
    }

    fn feature_branch(&self) -> &str {
        if let Some(branch) = &self.args.feature_branch {
            branch
        } else if let Some(branch) = &self.file.feature_branch {
            branch
        } else {
            &self.config.feature_branch
        }
    }

    fn version_pattern(&self) -> &str {
        if let Some(branch) = &self.args.version_pattern {
            branch
        } else if let Some(branch) = &self.file.version_pattern {
            branch
        } else {
            &self.config.version_pattern
        }
    }

    fn prerelease_tag(&self) -> &str {
        if let Some(branch) = &self.args.prerelease_tag {
            branch
        } else if let Some(branch) = &self.file.prerelease_tag {
            branch
        } else {
            &self.config.prerelease_tag
        }
    }

    fn verbose(&self) -> bool {
        self.args.verbose
    }
}
