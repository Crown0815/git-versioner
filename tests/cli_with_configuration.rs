mod common;

use crate::common::MAIN_BRANCH;
use anyhow::{Result, anyhow};
use common::{TestRepo, cli};
use git_versioner::config::ConfigurationFile;
use insta::with_settings;
use insta_cmd::assert_cmd_snapshot;
use rstest::{fixture, rstest};
use std::fs;
use std::path::PathBuf;
use std::process::Command;

const CUSTOM_MAIN_BRANCH: &str = "stem";
const DEFAULT_CONFIG: &str = ".git-versioner";

struct ConfiguredTestRepo {
    inner: TestRepo,
    cli_config: ConfigurationFile,
}

impl ConfiguredTestRepo {
    pub fn path(&self) -> &PathBuf {
        &self.inner.path
    }
    pub fn graph(&self) -> String {
        self.inner.graph()
    }

    pub fn create_config(&self, name: &str, extension: &str) -> Result<PathBuf> {
        let content = match extension {
            "toml" => toml::to_string(&self.cli_config)?,
            "yaml" => serde_yaml::to_string(&self.cli_config)?,
            &_ => return Err(anyhow!("Unsupported file extension {extension}")),
        };
        self.write(name, extension, content)
    }

    fn write(&self, filename: &str, extension: &str, content: String) -> Result<PathBuf> {
        let config_path = self.path().join(format!("{filename}.{extension}"));
        fs::write(&config_path, content)?;
        Ok(config_path)
    }
}

#[macro_export]
macro_rules! assert_configured_repo_cmd_snapshot {
    ($repo:expr, $config:expr, $cmd:expr) => {
        with_settings!(
            { description =>
                format!(
                    "Git Graph:\n    {}\nConfiguration ({}):\n    {}",
                    $repo.graph().replace("\n", &format!("\n{}", "    ")).trim_end_matches(' '),
                    $config.file_name().unwrap().to_string_lossy(),
                    fs::read_to_string(&$config).unwrap().replace("\n", &format!("\n{}", "    ")).trim_end_matches(' ')
                )
            },
            { assert_cmd_snapshot!($cmd); }
        );
    };
}

#[fixture]
pub fn repo(#[default(MAIN_BRANCH)] main: &str) -> ConfiguredTestRepo {
    let repo = TestRepo::initialize(main);
    let cli_config = ConfigurationFile::default();
    repo.commit("0.1.0-pre.1");

    ConfiguredTestRepo {
        inner: repo,
        cli_config,
    }
}

#[rstest]
fn test_that_toml_config_file_overrides_default_main_branch_pattern(
    #[with(CUSTOM_MAIN_BRANCH)] mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.main_branch = Some(format!("^{CUSTOM_MAIN_BRANCH}$"));
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();

    assert_configured_repo_cmd_snapshot!(repo, config_file, cli.current_dir(repo.path()));
}

#[rstest]
fn test_that_cli_argument_overrides_configuration_of_main_branch_pattern(
    #[with(CUSTOM_MAIN_BRANCH)] mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.main_branch = Some(format!("^{}$", "another_main_branch"));
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();

    assert_configured_repo_cmd_snapshot!(
        repo,
        config_file,
        cli.current_dir(repo.path())
            .args(["--main-branch", CUSTOM_MAIN_BRANCH])
    );
}

#[rstest]
fn test_that_toml_config_file_overrides_default_release_branch_pattern(
    mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.release_branch = Some("custom-release/(?<BranchName>.*)".to_string());
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();
    repo.inner.commit("0.1.0-pre.1");
    repo.inner.tag("v1.0.0");
    repo.inner.branch("custom-release/1.0.0");
    repo.inner.commit("1.0.1-pre.1");
    assert_configured_repo_cmd_snapshot!(repo, config_file, cli.current_dir(repo.path()));
}

#[rstest]
fn test_that_cli_argument_overrides_configuration_of_release_branch_pattern(
    mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.release_branch = Some("whatever-release/(?<BranchName>.*)".to_string());
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();
    repo.inner.commit("0.1.0-pre.1");
    repo.inner.tag("v1.0.0");
    repo.inner.branch("custom-release/1.0.0");
    repo.inner.commit("1.0.1-pre.1");
    assert_configured_repo_cmd_snapshot!(
        repo,
        config_file,
        cli.current_dir(repo.path())
            .args(["--release-branch", "custom-release/(?<BranchName>.*)"])
    );
}

#[rstest]
fn test_that_toml_config_file_overrides_default_feature_branch_pattern(
    mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.feature_branch = Some("my-feature/(?<BranchName>.*)".to_string());
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();
    repo.inner.commit("0.1.0-pre.1");
    repo.inner.branch("my-feature/feature");
    repo.inner.commit("0.1.0-feature.1");
    assert_configured_repo_cmd_snapshot!(repo, config_file, cli.current_dir(repo.path()));
}

#[rstest]
fn test_that_cli_argument_overrides_configuration_of_feature_branch_pattern(
    mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.feature_branch = Some("whatever-feature/(?<BranchName>.*)".to_string());
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();
    repo.inner.commit("0.1.0-pre.1");
    repo.inner.branch("my-feature/feature");
    repo.inner.commit("0.1.0-feature.1");
    assert_configured_repo_cmd_snapshot!(
        repo,
        config_file,
        cli.current_dir(repo.path())
            .args(["--feature-branch", "my-feature/(?<BranchName>.*)"])
    );
}

#[rstest]
fn test_that_toml_config_file_overrides_default_version_pattern(
    mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.version_pattern = Some("my/c(?<Version>.*)".to_string());
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();
    repo.inner.commit("0.1.0-pre.1");
    repo.inner.tag("my/c1.0.0");

    assert_configured_repo_cmd_snapshot!(repo, config_file, cli.current_dir(repo.path()));
}

#[rstest]
fn test_that_cli_argument_overrides_configuration_of_version_pattern(
    mut repo: ConfiguredTestRepo,
    #[values("toml", "yaml")] extension: &str,
    mut cli: Command,
) {
    repo.cli_config.version_pattern = Some("my/c(?<Version>.*)".to_string());
    let config_file = repo.create_config(DEFAULT_CONFIG, extension).unwrap();
    repo.inner.commit("0.1.0-pre.1");
    repo.inner.tag("my/v1.0.0");

    assert_configured_repo_cmd_snapshot!(
        repo,
        config_file,
        cli.current_dir(repo.path())
            .args(["--version-pattern", "my/v(?<Version>.*)"])
    );
}
