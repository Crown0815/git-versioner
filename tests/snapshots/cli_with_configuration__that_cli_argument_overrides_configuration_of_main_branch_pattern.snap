---
source: tests/cli_with_configuration.rs
description: "Git Graph:\n    * ##SHA## (HEAD -> stem) 0.1.0-pre.1\n\nConfiguration (.git-versioner.toml):\n    MainBranch = \"^another_main_branch$\"\n"
info:
  program: git-versioner
  args:
    - "--main-branch"
    - stem
---
success: true
exit_code: 0
----- stdout -----
{
  "Major": 0,
  "Minor": 1,
  "Patch": 0,
  "MajorMinorPatch": "0.1.0",
  "PreReleaseTag": "pre.1",
  "PreReleaseTagWithDash": "-pre.1",
  "PreReleaseLabel": "pre",
  "PreReleaseLabelWithDash": "-pre",
  "PreReleaseNumber": "1",
  "BuildMetadata": "",
  "SemVer": "0.1.0-pre.1",
  "AssemblySemVer": "0.1.0",
  "FullSemVer": "0.1.0-pre.1",
  "InformationalVersion": "0.1.0-pre.1",
  "BranchName": "stem",
  "EscapedBranchName": "stem"
}

----- stderr -----
